sycl/source/detail/device_image_impl.hpp

```diff
@@ -150,6 +160,35 @@ class device_image_impl {
   }
 
 private:
+  void updateSpecConstSymMap() {
+    if (MBinImage) {
+      const pi_device_binary_struct &RawImg = MBinImage->getRawData();
+      const auto &PropSetsBegin = RawImg.PropertySetsBegin;
+      const auto &PropSetsEnd = RawImg.PropertySetsEnd;
+      const auto &SpecConstMap = std::find_if(
+          PropSetsBegin, PropSetsEnd,
+          [](const _pi_device_binary_property_set_struct &Set) {
+            return strcmp(Set.Name, __SYCL_PI_PROPERTY_SET_SPEC_CONST_MAP) == 0;
+          });
+
+      if (SpecConstMap != PropSetsEnd) {
+        const auto &PropsBegin = SpecConstMap->PropertiesBegin;
+        const auto &PropsEnd = SpecConstMap->PropertiesEnd;
+
+        std::for_each(
+            PropsBegin, PropsEnd,
+            [&](const _pi_device_binary_property_struct &Prop) {
+              MSpecConstSymMap[Prop.Name] =
```

Comment:
If I understand correctly there can be several spec consts with the same name in case when spec const is an aggregate type. So, it seems we need to make sure a string maps to the first ID of such a type.
@AlexeySachkov Could you please confirm?

Reply:
That's correct. Composite specialization constant at SYCL level is being lowered into a set of scalar specialization constants at SPIR-V level. Please also note that since property is a byte-array, first 8 bytes of it are allocated for its size, you can find an example of how we handle it here: https://github.com/intel/llvm/blob/sycl/sycl/source/detail/program_impl.cpp#L547